component/src/main/java/io/siddhi/extension/execution/string/FillTemplateFunctionExtension.java

```diff
@@ -118,23 +124,45 @@
     @Override
     protected Object execute(Object[] objects, State state) {
         String sourceString = (String) objects[0];
+        Map<String, Object> valueMap = null;
         Matcher templateMatcher = templatePattern.matcher(sourceString);
         String match;
         Matcher indexMatcher;
-        int index;
-        while (templateMatcher.find()) {
-            match = templateMatcher.group(0);
-            indexMatcher = indexPattern.matcher(match);
-            if (indexMatcher.find()) {
-                index = Integer.parseInt(indexMatcher.group(0));
-                if (index < 1 || index >= objects.length) {
-                    throw new SiddhiAppRuntimeException("Index given for template elements "
-                            + "should be greater than 0 and less than '" + objects.length + "'. But found "
-                            + index + " in" + " the template '" + sourceString + "'.");
+        if (objects[1] instanceof HashMap) {
+            valueMap = (HashMap<String, Object>) objects[1];
+            String key = "";
+            while (templateMatcher.find()) {
+                match = templateMatcher.group(0);
+                indexMatcher = indexPattern.matcher(match);
+                if (indexMatcher.find()) {
+                    key = indexMatcher.group(0);
+                    if (key == null || key.equals("")) {
+                        throw new SiddhiAppRuntimeException("Key given for template elements "
+                                + "should be greater not null or empty. But found "
+                                + " in" + " the template '" + sourceString + "'.");
+                    }
+                    if (valueMap.get(key) != null) {
+                        sourceString = sourceString.replace(match, valueMap.get(key).toString());
```

Comment:
valueMap.get(key).toString() might not get string value in all cases. Will have to handle according to its type. Check and use String.valueOf(number) if it works for all cases.